fsmonitor.c

```diff
@@ -32,17 +46,27 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 	uint32_t ewah_size;
 	struct ewah_bitmap *fsmonitor_dirty;
 	int ret;
+	uint64_t tm;
+	struct strbuf last_update = STRBUF_INIT;
 
 	if (sz < sizeof(uint32_t) + sizeof(uint64_t) + sizeof(uint32_t))
 		return error("corrupt fsmonitor extension (too short)");
 
 	hdr_version = get_be32(index);
 	index += sizeof(uint32_t);
-	if (hdr_version != INDEX_EXTENSION_VERSION)
+	if (hdr_version == INDEX_EXTENSION_VERSION1) {
+		tm = get_be64(index);
+		strbuf_addf(&last_update, "%"PRIu64"", tm);
+		index += sizeof(uint64_t);
+	} else if (hdr_version == INDEX_EXTENSION_VERSION2) {
+		strbuf_addstr(&last_update, index);
+		index += last_update.len + 1;
+	} else {
+		strbuf_release(&last_update);
 		return error("bad fsmonitor version %d", hdr_version);
+	}
 
-	istate->fsmonitor_last_update = get_be64(index);
-	index += sizeof(uint64_t);
+	istate->fsmonitor_last_update = strbuf_detach(&last_update, NULL);
 
 	ewah_size = get_be32(index);
```

Comment:
Just checking:  Previously, the header was a nicely aligned { ui32, ui64, ui32 } thing.  Now with the token, we have no guards on the size of the middle field, right?  So the `ewah_size` field might be filled using unaligned data, since `index` might not be a multiple of 4 at this point.  Again, just checking.

Reply:
I was wondering about this.  Does the ewah bitmap need to be aligned?  If so I will need to do some work to make sure it is.

Reply:
not sure about ewah's themselves, you might ask @derrickstolee about that (he's played with them and I haven't).  I was mainly worried about an unaligned processor fault inside get_be32() on finicky processors.

Reply:
`get_be32()` is defined in two versions, one guarded by the unwieldy condition:

```c
#if !defined(NO_UNALIGNED_LOADS) && ( \
    defined(__i386__) || defined(__x86_64__) || \
    defined(_M_IX86) || defined(_M_X64) || \
    defined(__ppc__) || defined(__ppc64__) || \
    defined(__powerpc__) || defined(__powerpc64__) || \
    defined(__s390__) || defined(__s390x__))
```

So I think we're actually safe, we do not have to worry that the `get_be*()` function family might stumble over alignment issues: they don't.

Reply:
(see `compat/bswap.h`)

Reply:
thanks!